### PR TITLE
rewriting most of the asn1 init code in ruby

### PR DESCRIPTION
**Pull request by HoneyryderChuck**
This is an initial proposal towards the goal of having as much of the lib in ruby as possible. While most of the code must be in C, some of the code in C is essentially calling back to ruby, and writing "ruby in C", if that makes any sense.

Lemme know what you think. I understand there may also be some cognitive load from jumping from C to ruby and back, so there's certainly a trade-off.

### CODE REVIEW

File: lib/openssl/asn1.rb [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+      return tag if tag
+
+      sklass = klass.superclass
+
+      return unless sklass
+
+      take_default_tag(sklass)
+    end
+  end
+end
```
Comment (junaruga) [inline lib/openssl/asn1.rb:119]: I think it's better to add a link break at the end of the file.

File: ext/openssl/ossl_asn1.c [hunk 2]
```diff
@@ -654,18 +654,18 @@ ossl_asn1_class2sym(int tc)
...
+// {
+//     if(!SYMBOL_P(tag_class))
+// 	ossl_raise(eASN1Error, "invalid tag class");
+//     ossl_asn1_set_tag(self, tag);
+//     ossl_asn1_set_value(self, value);
+//     ossl_asn1_set_tag_class(self, tag_class);
+//     ossl_asn1_set_indefinite_length(self, Qfalse);
+
+//     return self;
+// }
```
Comment (junaruga) [inline ext/openssl/ossl_asn1.c:668]: Why did you comment out by `//` rather than removing the parts?

Comment (HoneyryderChuck) [reply ext/openssl/ossl_asn1.c:668]: This is an early draft, just wanted to get an opinion about the approach before moving forward (still some more work to be done).

File: lib/openssl/asn1.rb (hunk 1 above)
Comment (HoneyryderChuck) [reply lib/openssl/asn1.rb]: done

File: ext/openssl/ossl_asn1.c (hunk 2 above)
Comment (HoneyryderChuck) [reply ext/openssl/ossl_asn1.c]: done

**Comment by rhenium** [comment]
I'm all for it!

Apart from methods for accessing OpenSSL's OID registry (`OpenSSL::ASN1::ObjectID.register`, `#sn`, `#ln`, and `#oid`), `OpenSSL::ASN1` could have been entirely written in Ruby, and that would definitely make maintenance easier.

It would be very nice if we can port the BER decoding code to Ruby.

**Comment by HoneyryderChuck** [comment]
@rhenium thx for the feedback!

Some follow-up where I need your input as well:

If you see the ruby code, you'll find the `initialize` method copy-pasted both for `Primitive` and `Constructive`, which is a direct port from the deleted C code, where both classes were sharing the same underlying impl in C. I'd like to address, and I can see the following options:

* Create base class for both, inheriting from `ASN1Data`, which implements common `initialize`;
* Create module with `initialize`, inject on both classes;
* Create `__initialize_with..` method in `ASN1Data` class, call for both on `initialize`;

Essentially, the first 2 options will require a potential breaking(?) change in the ancestor chain. Not sure how you feel about that. The third option doesn't do that, at the cost of adding an additional "hidden" method. All of the above will add smth new to document (or maybe we can `:nodoc:` all of them?). Which one do you think suits this the best?

> It would be very nice if we can port the BER decoding code to Ruby.

You mean `OpenSSL::ASN1.decode`? I can try, but AFAIR there was some dependency on some openssl function calls, so don't know how feasible will that be.

**Comment by rhenium** [comment]
I'm sorry for the delay.

> If you see the ruby code, you'll find the `initialize` method copy-pasted both for `Primitive` and `Constructive`, which is a direct port from the deleted C code, where both classes were sharing the same underlying impl in C. I'd like to address, and I can see the following options:
> 
>     * Create base class for both, inheriting from `ASN1Data`, which implements common `initialize`;
> 
>     * Create module with `initialize`, inject on both classes;
> 
>     * Create `__initialize_with..` method in `ASN1Data` class, call for both on `initialize`;
> 
> 
> Essentially, the first 2 options will require a potential breaking(?) change in the ancestor chain. Not sure how you feel about that. The third option doesn't do that, at the cost of adding an additional "hidden" method. All of the above will add smth new to document (or maybe we can `:nodoc:` all of them?). Which one do you think suits this the best?

Adding a class or module in the ancestor chain is normally a compatible change. Private/hidden method also seems fine. At the same time, the current diff doesn't seem to contain too much amount of code duplication to me. Please use whatever you're comfortable with.

**Comment by rhenium** [comment]
> > It would be very nice if we can port the BER decoding code to Ruby.
> 
> You mean `OpenSSL::ASN1.decode`? I can try, but AFAIR there was some dependency on some openssl function calls, so don't know how feasible will that be.

It uses very low-level OpenSSL functions such as `ASN1_get_object()` or `d2i_ASN1_INTEGER()` which I hope aren't too difficult to replace. However this is a completely separate topic and not a blocker for this PR>

**Comment by HoneyryderChuck** [comment]
@rhenium done.

File: ext/openssl/ossl_asn1.c [hunk 3]
```diff
@@ -795,23 +765,22 @@ int_ossl_asn1_decode0_prim(unsigned char **pp, long length, long hlen, int tag,
...
-	args[1] = INT2NUM(tag);
-	args[2] = Qnil;
-	args[3] = tc;
-	asn1data = rb_obj_alloc(klass);
-	ossl_asn1_initialize(4, args, asn1data);
+	if (tag == V_ASN1_EOC)
+		asn1data = rb_funcall(cASN1EndOfContent, rb_intern("new"), 0);
+	else {
+		VALUE args[4] = { value, INT2NUM(tag), Qnil, tc };
+		asn1data = rb_funcall3(klass, rb_intern("new"), 4, args);
```
Comment (rhenium) [inline ext/openssl/ossl_asn1.c:772]: Please use the new name `rb_funcallv_public()`.

File: ext/openssl/ossl_asn1.c [hunk 4]
```diff
@@ -842,20 +811,22 @@ int_ossl_asn1_decode0_cons(unsigned char **pp, long max_len, long length,
...
-	else
-	    asn1data = rb_obj_alloc(cASN1Constructive);
-	args[0] = ary;
-	args[1] = INT2NUM(tag);
-	args[2] = Qnil;
-	args[3] = tc;
-	ossl_asn1_initialize(4, args, asn1data);
-    }
+		if (tag == V_ASN1_EOC)
+			asn1data = rb_funcall(cASN1EndOfContent, rb_intern("new"), 0);
```
Comment (rhenium) [inline ext/openssl/ossl_asn1.c:815]: This should be `cASN1Constructive`.

Only sequence, set, and string types may use constructed encoding, so this is actually an error, but let's not change the current behavior in this PR.

File: lib/openssl/asn1.rb [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+      # as parameter _asn1_. If no block is given, an enumerator is returned
+      # instead.
+      #
+      # == Example
+      #   asn1_ary.each do |asn1|
+      #     puts asn1
+      #   end
+      #
+      def each(&blk)
+        @value.each(&blk)
```
Comment (rhenium) [inline lib/openssl/asn1.rb:149]: This method should return `self`.

File: ext/openssl/ossl_asn1.c [hunk 6]
```diff
@@ -795,20 +765,19 @@ int_ossl_asn1_decode0_prim(unsigned char **pp, long length, long hlen, int tag,
...
-	args[1] = INT2NUM(tag);
-	args[2] = Qnil;
-	args[3] = tc;
-	asn1data = rb_obj_alloc(klass);
-	ossl_asn1_initialize(4, args, asn1data);
+    if (tag == V_ASN1_EOC)
+        asn1data = rb_funcall(cASN1EndOfContent, rb_intern("new"), 0);
+    else {
+	    VALUE args[4] = { value, INT2NUM(tag), Qnil, tc };
+        asn1data = rb_funcall3(klass, rb_intern("new"), 4, args);
```
[the remark this answers is not in the document]
Comment (rhenium) [reply ext/openssl/ossl_asn1.c]: This is not fixed yet. A hard tab has an equal width to 8 spaces.

**Comment by rhenium** [comment]
GitHub doesn't let me add comments to the collapsed part of the diff, but we can cleanup some definitions in the C file:

 - `ossl_asn1_set_*()` macros except for `ossl_asn1_set_indefinite_length()` are no longer used and should be removed.
 - `id_each` can also be removed.

**Comment by rhenium** [comment]
```
Run bundle exec rake compile
rake aborted!
NameError: undefined method `indefinite_length=' for class `OpenSSL::ASN1::Primitive'
/home/runner/work/openssl/openssl/lib/openssl/asn1.rb:128:in `undef_method'
/home/runner/work/openssl/openssl/lib/openssl/asn1.rb:128:in `<class:Primitive>'
/home/runner/work/openssl/openssl/lib/openssl/asn1.rb:125:in `<module:ASN1>'
/home/runner/work/openssl/openssl/lib/openssl/asn1.rb:12:in `<module:OpenSSL>'
/home/runner/work/openssl/openssl/lib/openssl/asn1.rb:11:in `<top (required)>'
/home/runner/work/openssl/openssl/lib/openssl.rb:16:in `require_relative'
/home/runner/work/openssl/openssl/lib/openssl.rb:16:in `<top (required)>'
/home/runner/work/openssl/openssl/vendor/bundle/ruby/2.7.0/gems/rake-compiler-1.2.7/lib/rake/extensiontask.rb:5:in `require'
/home/runner/work/openssl/openssl/vendor/bundle/ruby/2.7.0/gems/rake-compiler-1.2.7/lib/rake/extensiontask.rb:5:in `<top (required)>'
/home/runner/work/openssl/openssl/Rakefile:6:in `require'
/home/runner/work/openssl/openssl/Rakefile:6:in `<top (required)>'
/home/runner/work/openssl/openssl/vendor/bundle/ruby/2.7.0/gems/rake-13.2.1/exe/rake:27:in `<top (required)>'
/opt/hostedtoolcache/Ruby/2.7.8/x64/bin/bundle:23:in `load'
/opt/hostedtoolcache/Ruby/2.7.8/x64/bin/bundle:23:in `<main>'
(See full trace by running task with --trace)
Error: Process completed with exit code 1.
```

GitHub Actions is failing with this error with Ruby 2.7, but I'm not sure why this occurs. I can't reproduce it locally.

File: ext/openssl/ossl_asn1.c (hunk 4 above)
Comment (HoneyryderChuck) [reply ext/openssl/ossl_asn1.c:815]: I'm removing this conditional, as it's not porting anything from main branch, and I forgot the original context.

**Comment by HoneyryderChuck** [comment]
> GitHub Actions is failing with this error with Ruby 2.7, but I'm not sure why this occurs.

This was a bug, which was fixed in ruby 3: attr_accessor does not register `#a` and `#a=` as instance methods, so `undef_method` call does not find them.

I've resorted to not undeffing them in ruby 2.7 . WDYT? I'm assuming you'll be dropping 2.7 one of these days, and that's less work than figuring out a more invasive patch than this one, but I don't know if there are other implications.

**Comment by rhenium** [comment]
I haven't investigated why it raises an exception in Ruby 2.7 only, but the root cause of this particular error in `bundle exec rake compile` is that it's loading the Ruby part from local `lib/` and the C part from Ruby's default gem. This was happening with all Ruby versions. #768 should fix it.

File: ext/openssl/ossl_asn1.c [hunk 7]
```diff
@@ -795,23 +760,22 @@ int_ossl_asn1_decode0_prim(unsigned char **pp, long length, long hlen, int tag,
...
 	}
-    }
-    else {
-	asn1data = rb_obj_alloc(cASN1Data);
-	ossl_asn1data_initialize(asn1data, value, INT2NUM(tag), tc);
-    }
+	}
+	else {
+		VALUE args[3] = { value, INT2NUM(tag), tc };
+		asn1data = rb_funcall3(cASN1Data, rb_intern("new"), 3, args);
```
Comment (rhenium) [inline ext/openssl/ossl_asn1.c:775]: Please use `rb_funcallv_public()` everywhere.

File: lib/openssl/asn1.rb [hunk 8]
```diff
@@ -0,0 +1,190 @@
+# frozen_string_literal: true
+#--
+#
+# = Ruby-space definitions that completes C-space funcs for BN
```
Comment (rhenium) [inline lib/openssl/asn1.rb:4]: BN -> ASN.1

File: lib/openssl/asn1.rb [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+      end
+    end
+
+    class Primitive < ASN1Data
+      include TaggedASN1Data
+
+      unless RUBY_VERSION < "3.0.0"
+        undef_method :indefinite_length=
+        undef_method :infinite_length=
+      end
```
Comment (rhenium) [inline lib/openssl/asn1.rb:131]: This version guard should be unnecessary now.

**Review by rhenium — CHANGES_REQUESTED** [review]
Please check the indentation settings in the C file.

Apart from these style issues, this looks good to me!

**Comment by HoneyryderChuck** [comment]
Done.

> Please check the indentation settings in the C file.

I'm struggling a bit with this one. I've changed them back to tabs, as most of the rest of the file is. Isn't it correct?

**Comment by rhenium** [comment]
> > Please check the indentation settings in the C file.
> 
> I'm struggling a bit with this one. I've changed them back to tabs, as most of the rest of the file is. Isn't it correct?

Please use 4 spaces for indentation. You can see the changed lines currently have an inconsistent indentation here: https://github.com/ruby/openssl/blob/bbacdfa19370d98212ecf2aae0eb7a9b544966e2/ext/openssl/ossl_asn1.c

Historically 8 consecutive spaces of indentation was replaced by a hard tab, which is why this file contains hard tabs, but this step is not needed anymore.

File: ext/openssl/ossl_asn1.c (hunk 7 above)
Comment (rhenium) [reply ext/openssl/ossl_asn1.c]: Could you update other `rb_funcall3()` usage too?

**Comment by HoneyryderChuck** [comment]
@rhenium hopefully it's better now.

**Review by rhenium — CHANGES_REQUESTED** [review]

**Comment by rhenium** [comment]
> @rhenium hopefully it's better now.

Yes, it looks good! Thanks.

File: ext/openssl/ossl_asn1.c [hunk 10]
```diff
@@ -795,20 +760,19 @@ int_ossl_asn1_decode0_prim(unsigned char **pp, long length, long hlen, int tag,
...
+	}
 	if(tag == V_ASN1_BIT_STRING){
 	    rb_ivar_set(asn1data, sivUNUSED_BITS, LONG2NUM(flag));
 	}
     }
     else {
-	asn1data = rb_obj_alloc(cASN1Data);
-	ossl_asn1data_initialize(asn1data, value, INT2NUM(tag), tc);
+        VALUE args[3] = { value, INT2NUM(tag), tc };
+        asn1data = rb_funcall3(cASN1Data, rb_intern("new"), 3, args);
```
[the remark this answers is not in the document]
Comment (HoneyryderChuck) [reply ext/openssl/ossl_asn1.c]: done

**Review by rhenium — APPROVED** [review]

**Comment by rhenium** [comment]
Merged now. Thank you!

**Pull request merged**